walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,113 @@
+    List<TimeRange> invalidTimes = new ArrayList<>();  // keeps track of invalid ranges
+    Collection<String> attendees = request.getAttendees();
+
+    int duration = (int) request.getDuration();
+
+    // if length of meeting is longer than a whole day, return empty list
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+        return validTimes;
+    }
+
+    // if no conflicting events
+    if (events.isEmpty()) {
+        validTimes.add(TimeRange.WHOLE_DAY);
+        return validTimes;
+    }
+
+    // Add to invalid times collection
+    for (Event event : events) {
+        for(String attendee : attendees) {
+            if(event.getAttendees().contains(attendee)) {
+                invalidTimes.add(event.getWhen());
+            }
+        }
+    }
+
+    // if no events for attendees
+    if (invalidTimes.isEmpty()) {
+        validTimes.add(TimeRange.WHOLE_DAY);
+        return validTimes;
+    }
+
+    invalidTimes = mergeOverlap(invalidTimes);
+
+    return findValidTimes(invalidTimes, duration);
+  }
+
+  /*
+  * Merges overlapping/nested events into larger ranges
+  */
+  private List<TimeRange> mergeOverlap(List<TimeRange> invalidTimes) {
+    // sort in order for ease in finding overlapping events
+    List<TimeRange> mergedTimes = new ArrayList<>(invalidTimes);
+    Collections.sort(mergedTimes, TimeRange.ORDER_BY_START);  // sort by start time
+
+    // finds overlapping invalid events 
+    for (int i = 0; i < mergedTimes.size() - 1; i++) {
+        TimeRange eventOne = mergedTimes.get(i);
+        TimeRange eventTwo = mergedTimes.get(i+1);
+        // choose longer event in case of nesting
+        if (eventOne.contains(eventTwo)) {
+            mergedTimes.remove(eventTwo);
+            continue;
+        }
+        if (eventTwo.contains(eventOne)) {
+            mergedTimes.remove(eventOne);
+            continue;
+        }
+        if (!eventOne.overlaps(mergedTimes.get(i+1))) {
+            continue;
+        }
+        // if overlap, create new range
+        TimeRange newRange = TimeRange.fromStartEnd(eventOne.start(), eventTwo.end(), false);
+        mergedTimes.remove(eventOne);
+        mergedTimes.remove(eventTwo);
+        mergedTimes.add(i, newRange);
+    }
+    return mergedTimes;
+  }
+
+  /*
+  * Find valid times from invalid events/times
+  */
+  private List<TimeRange> findValidTimes(List<TimeRange> invalidTimes, int duration) {
+    List<TimeRange> validTimes = new ArrayList<>();
+    boolean first = false;  // first event of the day
+    int start = TimeRange.START_OF_DAY;  // current start of meeting
+
+    for (int i = 0; i < invalidTimes.size(); i++) {
+        TimeRange invalid = invalidTimes.get(i);
+
+        // if first event at beginning of day, move start to end of event
+        if (invalid.start() == TimeRange.START_OF_DAY && first) {
+            start = invalid.end();
+            first = false;
+            continue;
+        }
```

Comment:
Seems to me that you don't actually need this.

Reply:
You're right! I had this originally because it's leftover from an earlier implementation I did.